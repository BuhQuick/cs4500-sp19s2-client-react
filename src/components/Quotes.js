import React from 'react'
const Quotes = ({quotes, quote,
  selectQuote, updateForm, createQuote,
  updateQuote, deleteQuote}) =>
<div>
    <h3>Quotes {quotes.length}</h3>
    <table className="table">
        <tbody>
        <tr>
            <td>
                <input
                    className="job-title-fld"
                    onChange={e => updateForm(e)}
                    value={quote.jobTitle}></input>
                <button
                    className="create-quote-btn"
                    onClick={createQuote}>Add</button>
                <button
                    className="update-quote-btn"
                    onClick={updateQuote}>Update</button>
            </td>
        </tr>
        {
            quotes
                .map(quote =>
                    <tr className="quote-row"
                        key={quote.id}>
                        <td className="job-title">
                            {quote.jobTitle}
                        </td>
                        <td>
                            <button
                                className="delete-quote-btn"
                                onClick={() => deleteQuote(quote.id)}>
                            Delete
                        </button>
                        <button
                            className="select-quote-btn"
                            onClick={() => selectQuote(quote.id)}>
                            Select
                        </button>
                        </td>
                    </tr>
                )
        }
        </tbody>
    </table>
</div>

export default Quotes
