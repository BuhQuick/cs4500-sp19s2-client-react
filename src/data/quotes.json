[
  {"id":1,"jobTitle":"J1","jobDescription":"D1","price":123},
  {"id":2,"jobTitle":"J2","jobDescription":"D2","price":234}
]