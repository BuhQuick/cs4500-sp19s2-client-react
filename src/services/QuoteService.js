export default class QuoteService {
    static instance = null;
    static getInstance() {
        if(QuoteService.instance === null) {
            QuoteService.instance = new QuoteService()
        }
        return this.instance
    }
    findQuoteById = quoteId =>
        fetch(`http://localhost:8080/api/quotes/${quoteId}`)
            .then(response => response.json())
    findAllQuotes = () =>
        fetch("http://localhost:8080/api/quotes")
            .then(response => response.json())
    createQuote = quote => {
        delete quote.id;
        return fetch(`http://localhost:8080/api/quotes`, {
            method: 'post',
            body: JSON.stringify(quote),
            headers: {
                'content-type': 'application/json'
            }
        })
            .then(response => response.json())
    }
    updateQuote = quote =>
        fetch(`http://localhost:8080/api/quotes/${quote.id}`, {
            method: 'put',
            body: JSON.stringify(quote),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())
    deleteQuote = id =>
        fetch(`http://localhost:8080/api/quotes/${id}`, {
            method: 'delete'
        })

}
