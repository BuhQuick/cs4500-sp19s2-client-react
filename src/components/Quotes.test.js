import React from 'react'
import Quotes from './Quotes'
import TestRenderer from 'react-test-renderer';
import quotes from '../data/quotes.json'
import QuoteService from '../services/QuoteService'
const quoteService = QuoteService.getInstance()
import '../services/QuoteService.mock'

test('create and delete quotes', (done) => {
    const createQuote = () => {
        quotes.push({id: quotes.length+1, jobTitle:"J123-" + quotes.length+1})

        let tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot()

        done()
    }

    const deleteQuote = quoteId => {
        quotes.splice(0, 1)

        let tree = testRenderer.toJSON();
        expect(tree).toMatchSnapshot()

        done()
    }

    const testRenderer = TestRenderer.create(
        <Quotes
            quote={quotes[0]}
            quotes={quotes}
            createQuote={createQuote}
            deleteQuote={deleteQuote}
        />)
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot()

    const testInstance = testRenderer.root

    const createQuoteBtn = testInstance.findByProps({className: 'create-quote-btn'})
    createQuoteBtn.props.onClick()
    createQuoteBtn.props.onClick()

    const deleteQuoteBtns = testInstance.findAllByProps({className: 'delete-quote-btn'})
    deleteQuoteBtns[1].props.onClick()
})

test('quote list renders correctly', () => {
    const testRenderer = TestRenderer.create(
        <Quotes
            quote={quotes[0]}
            quotes={quotes}/>)
    let tree = testRenderer.toJSON()
    expect(tree).toMatchSnapshot()

    const testInstance = testRenderer.root

    const jobTitleFld    = testInstance.findByProps({className: 'job-title-fld'})
    const createQuoteBtn = testInstance.findAllByProps({className: 'create-quote-btn'})
    const updateQuoteBtn = testInstance.findAllByProps({className: 'update-quote-btn'})
    const quoteRows = testInstance.findAllByProps({className: 'quote-row'})
    const jobTitles = testInstance.findAllByProps({className: 'job-title'})
    const deleteQuoteBtns = testInstance.findAllByProps({className: 'delete-quote-btn'})
    const selectQuoteBtns = testInstance.findAllByProps({className: 'select-quote-btn'})

    expect(jobTitleFld.props.value).toBe('J2')
    expect(createQuoteBtn.length).toBe(1)
    expect(updateQuoteBtn.length).toBe(1)
    expect(quoteRows.length).toBe(3)
    expect(jobTitles.length).toBe(3)
    expect(deleteQuoteBtns.length).toBe(3)
    expect(selectQuoteBtns.length).toBe(3)

})

test('render all quotes from services', () => {
    quoteService
        .findAllQuotes()
        .then(quotes => {
            const testRenderer = TestRenderer.create(
                <Quotes
                    quote={quotes[0]}
                    quotes={quotes}/>)
            let tree = testRenderer.toJSON()
            expect(tree).toMatchSnapshot()

            const testInstance = testRenderer.root

            const jobTitleFld    = testInstance.findByProps({className: 'job-title-fld'})
            const createQuoteBtn = testInstance.findAllByProps({className: 'create-quote-btn'})
            const updateQuoteBtn = testInstance.findAllByProps({className: 'update-quote-btn'})
            const quoteRows = testInstance.findAllByProps({className: 'quote-row'})
            const jobTitles = testInstance.findAllByProps({className: 'job-title'})
            const deleteQuoteBtns = testInstance.findAllByProps({className: 'delete-quote-btn'})
            const selectQuoteBtns = testInstance.findAllByProps({className: 'select-quote-btn'})

            expect(jobTitleFld.props.value).toBe('J2')
            expect(createQuoteBtn.length).toBe(1)
            expect(updateQuoteBtn.length).toBe(1)
            expect(quoteRows.length).toBe(3)
            expect(jobTitles.length).toBe(3)
            expect(deleteQuoteBtns.length).toBe(3)
            expect(selectQuoteBtns.length).toBe(3)

        })
})

