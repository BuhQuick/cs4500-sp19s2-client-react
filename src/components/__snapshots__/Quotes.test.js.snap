// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create and delete quotes 1`] = `
<div>
  <h3>
    Quotes 
    2
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="job-title-fld"
            onChange={[Function]}
            value="J1"
          />
          <button
            className="create-quote-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="update-quote-btn"
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J1
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J2
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`create and delete quotes 2`] = `
<div>
  <h3>
    Quotes 
    2
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="job-title-fld"
            onChange={[Function]}
            value="J1"
          />
          <button
            className="create-quote-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="update-quote-btn"
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J1
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J2
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`create and delete quotes 3`] = `
<div>
  <h3>
    Quotes 
    2
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="job-title-fld"
            onChange={[Function]}
            value="J1"
          />
          <button
            className="create-quote-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="update-quote-btn"
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J1
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J2
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`create and delete quotes 4`] = `
<div>
  <h3>
    Quotes 
    2
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="job-title-fld"
            onChange={[Function]}
            value="J1"
          />
          <button
            className="create-quote-btn"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            className="update-quote-btn"
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J1
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J2
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`quote list renders correctly 1`] = `
<div>
  <h3>
    Quotes 
    3
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="job-title-fld"
            onChange={[Function]}
            value="J2"
          />
          <button
            className="create-quote-btn"
          >
            Add
          </button>
          <button
            className="update-quote-btn"
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J2
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J123-21
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J123-31
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`render all quotes from services 1`] = `
<div>
  <h3>
    Quotes 
    3
  </h3>
  <table
    className="table"
  >
    <tbody>
      <tr>
        <td>
          <input
            className="job-title-fld"
            onChange={[Function]}
            value="J2"
          />
          <button
            className="create-quote-btn"
          >
            Add
          </button>
          <button
            className="update-quote-btn"
          >
            Update
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J2
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J123-21
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
      <tr
        className="quote-row"
      >
        <td
          className="job-title"
        >
          J123-31
        </td>
        <td>
          <button
            className="delete-quote-btn"
            onClick={[Function]}
          >
            Delete
          </button>
          <button
            className="select-quote-btn"
            onClick={[Function]}
          >
            Select
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
